File: tests/algorithms/x/OHP/strings/camel_case_to_snake_case.php
<?php
error_reporting(E_ALL & ~E_DEPRECATED);
ini_set('memory_limit', '-1');
$now_seed = 0;
$now_seeded = false;
$s = getenv('MOCHI_NOW_SEED');
if ($s !== false && $s !== '') {
    $now_seed = intval($s);
    $now_seeded = true;
}
function _now() {
    global $now_seed, $now_seeded;
    if ($now_seeded) {
        $now_seed = ($now_seed * 1664525 + 1013904223) % 2147483647;
        return $now_seed;
    }
    return hrtime(true);
}
$__start_mem = memory_get_usage();
$__start = _now();
  $LOWER = 'abcdefghijklmnopqrstuvwxyz';
  $UPPER = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ';
  $DIGITS = '0123456789';
  function is_lower($ch) {
  global $DIGITS, $LOWER, $UPPER;
  $i = 0;
  while ($i < strlen($LOWER)) {
  if (substr($LOWER, $i, $i + 1 - $i) == $ch) {
  return true;
}
  $i = $i + 1;
};
  return false;
};
  function is_upper($ch) {
  global $DIGITS, $LOWER, $UPPER;
  $i = 0;
  while ($i < strlen($UPPER)) {
  if (substr($UPPER, $i, $i + 1 - $i) == $ch) {
  return true;
}
  $i = $i + 1;
};
  return false;
};
  function is_digit($ch) {
  global $DIGITS, $LOWER, $UPPER;
  $i = 0;
  while ($i < strlen($DIGITS)) {
  if (substr($DIGITS, $i, $i + 1 - $i) == $ch) {
  return true;
}
  $i = $i + 1;
};
  return false;
};
  function is_alpha($ch) {
  global $DIGITS, $LOWER, $UPPER;
  if (is_lower($ch)) {
  return true;
}
  if (is_upper($ch)) {
  return true;
}
  return false;
};
  function is_alnum($ch) {
  global $DIGITS, $LOWER, $UPPER;
  if (is_alpha($ch)) {
  return true;
}
  if (is_digit($ch)) {
  return true;
}
  return false;
};
  function to_lower($ch) {
  global $DIGITS, $LOWER, $UPPER;
  $i = 0;
  while ($i < strlen($UPPER)) {
  if (substr($UPPER, $i, $i + 1 - $i) == $ch) {
  return substr($LOWER, $i, $i + 1 - $i);
}
  $i = $i + 1;
};
  return $ch;
};
  function camel_to_snake_case($input_str) {
  global $DIGITS, $LOWER, $UPPER;
  $snake_str = '';
  $i = 0;
  $prev_is_digit = false;
  $prev_is_alpha = false;
  while ($i < strlen($input_str)) {
  $ch = substr($input_str, $i, $i + 1 - $i);
  if (is_upper($ch)) {
  $snake_str = $snake_str . '_' . to_lower($ch);
} else {
  if ($prev_is_digit && is_lower($ch)) {
  $snake_str = $snake_str . '_' . $ch;
} else {
  if ($prev_is_alpha && is_digit($ch)) {
  $snake_str = $snake_str . '_' . $ch;
} else {
  if (!is_alnum($ch)) {
  $snake_str = $snake_str . '_';
} else {
  $snake_str = $snake_str . $ch;
};
};
};
}
  $prev_is_digit = is_digit($ch);
  $prev_is_alpha = is_alpha($ch);
  $i = $i + 1;
};
  if (strlen($snake_str) > 0 && substr($snake_str, 0, 0 + 1) == '_') {
  $snake_str = substr($snake_str, 1, strlen($snake_str) - 1);
}
  return $snake_str;
};
  function main() {
  global $DIGITS, $LOWER, $UPPER;
  echo rtrim(camel_to_snake_case('someRandomString')), PHP_EOL;
  echo rtrim(camel_to_snake_case('SomeRandomStr#ng')), PHP_EOL;
  echo rtrim(camel_to_snake_case('123SomeRandom123String123')), PHP_EOL;
};
  main();
$__end = _now();
$__end_mem = memory_get_peak_usage(true);
$__duration = max(1, intdiv($__end - $__start, 1000));
$__mem_diff = max(0, $__end_mem - $__start_mem);
$__bench = ["duration_us" => $__duration, "memory_bytes" => $__mem_diff, "name" => "main"];
$__j = json_encode($__bench, 128);
$__j = str_replace("    ", "  ", $__j);
echo $__j, PHP_EOL;

File: tests/algorithms/transpiler/FS/strings/camel_case_to_snake_case.fs
// Generated 2025-08-11 17:23 +0700

exception Return
let mutable _nowSeed:int64 = 0L
let mutable _nowSeeded = false
let _initNow () =
    let s = System.Environment.GetEnvironmentVariable("MOCHI_NOW_SEED")
    if System.String.IsNullOrEmpty(s) |> not then
        match System.Int32.TryParse(s) with
        | true, v ->
            _nowSeed <- int64 v
            _nowSeeded <- true
        | _ -> ()
let _now () =
    if _nowSeeded then
        _nowSeed <- (_nowSeed * 1664525L + 1013904223L) % 2147483647L
        int _nowSeed
    else
        int (System.DateTime.UtcNow.Ticks % 2147483647L)

_initNow()
let _substring (s:string) (start:int) (finish:int) =
    let len = String.length s
    let mutable st = if start < 0 then len + start else start
    let mutable en = if finish < 0 then len + finish else finish
    if st < 0 then st <- 0
    if st > len then st <- len
    if en > len then en <- len
    if st > en then st <- en
    s.Substring(st, en - st)

let _idx (arr:'a array) (i:int) : 'a =
    if not (obj.ReferenceEquals(arr, null)) && i >= 0 && i < arr.Length then arr.[i] else Unchecked.defaultof<'a>
let LOWER: string = "abcdefghijklmnopqrstuvwxyz"
let UPPER: string = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
let DIGITS: string = "0123456789"
let rec is_lower (ch: string) =
    let mutable __ret : bool = Unchecked.defaultof<bool>
    let mutable ch = ch
    try
        let mutable i: int = 0
        while i < (String.length (LOWER)) do
            if (string (LOWER.[i])) = ch then
                __ret <- true
                raise Return
            i <- i + 1
        __ret <- false
        raise Return
        __ret
    with
        | Return -> __ret
and is_upper (ch: string) =
    let mutable __ret : bool = Unchecked.defaultof<bool>
    let mutable ch = ch
    try
        let mutable i: int = 0
        while i < (String.length (UPPER)) do
            if (string (UPPER.[i])) = ch then
                __ret <- true
                raise Return
            i <- i + 1
        __ret <- false
        raise Return
        __ret
    with
        | Return -> __ret
and is_digit (ch: string) =
    let mutable __ret : bool = Unchecked.defaultof<bool>
    let mutable ch = ch
    try
        let mutable i: int = 0
        while i < (String.length (DIGITS)) do
            if (string (DIGITS.[i])) = ch then
                __ret <- true
                raise Return
            i <- i + 1
        __ret <- false
        raise Return
        __ret
    with
        | Return -> __ret
and is_alpha (ch: string) =
    let mutable __ret : bool = Unchecked.defaultof<bool>
    let mutable ch = ch
    try
        if is_lower (ch) then
            __ret <- true
            raise Return
        if is_upper (ch) then
            __ret <- true
            raise Return
        __ret <- false
        raise Return
        __ret
    with
        | Return -> __ret
and is_alnum (ch: string) =
    let mutable __ret : bool = Unchecked.defaultof<bool>
    let mutable ch = ch
    try
        if is_alpha (ch) then
            __ret <- true
            raise Return
        if is_digit (ch) then
            __ret <- true
            raise Return
        __ret <- false
        raise Return
        __ret
    with
        | Return -> __ret
and to_lower (ch: string) =
    let mutable __ret : string = Unchecked.defaultof<string>
    let mutable ch = ch
    try
        let mutable i: int = 0
        while i < (String.length (UPPER)) do
            if (string (UPPER.[i])) = ch then
                __ret <- string (LOWER.[i])
                raise Return
            i <- i + 1
        __ret <- ch
        raise Return
        __ret
    with
        | Return -> __ret
and camel_to_snake_case (input_str: string) =
    let mutable __ret : string = Unchecked.defaultof<string>
    let mutable input_str = input_str
    try
        let mutable snake_str: string = ""
        let mutable i: int = 0
        let mutable prev_is_digit: bool = false
        let mutable prev_is_alpha: bool = false
        while i < (String.length (input_str)) do
            let ch: string = string (input_str.[i])
            if is_upper (ch) then
                snake_str <- (snake_str + "_") + (to_lower (ch))
            else
                if prev_is_digit && (is_lower (ch)) then
                    snake_str <- (snake_str + "_") + ch
                else
                    if prev_is_alpha && (is_digit (ch)) then
                        snake_str <- (snake_str + "_") + ch
                    else
                        if not (is_alnum (ch)) then
                            snake_str <- snake_str + "_"
                        else
                            snake_str <- snake_str + ch
            prev_is_digit <- is_digit (ch)
            prev_is_alpha <- is_alpha (ch)
            i <- i + 1
        if ((String.length (snake_str)) > 0) && ((string (snake_str.[0])) = "_") then
            snake_str <- _substring snake_str (1) (String.length (snake_str))
        __ret <- snake_str
        raise Return
        __ret
    with
        | Return -> __ret
and main () =
    let mutable __ret : unit = Unchecked.defaultof<unit>
    try
        let __bench_start = _now()
        let __mem_start = System.GC.GetTotalMemory(true)
        printfn "%s" (camel_to_snake_case ("someRandomString"))
        printfn "%s" (camel_to_snake_case ("SomeRandomStr#ng"))
        printfn "%s" (camel_to_snake_case ("123SomeRandom123String123"))
        let __bench_end = _now()
        let __mem_end = System.GC.GetTotalMemory(true)
        printfn "{\n  \"duration_us\": %d,\n  \"memory_bytes\": %d,\n  \"name\": \"main\"\n}" ((__bench_end - __bench_start) / 1000) (__mem_end - __mem_start)

        __ret
    with
        | Return -> __ret
main()
